### PR TITLE
PEPPER-749 kit type name fix

### PR DESCRIPTION
**Pull request by ssettipalli**
https://broadworkbench.atlassian.net/browse/PEPPER-749
As per my investigation  looks like issue seem to be narrowed down to reactivated kits (from deactivated kits) .
Solution:
During the final scan, query the kit type name and update in elastic kitRequestShipping object.
Once these changes are merged to dev, need to test other scenarios (non re-activated kits)  as well.
Code fix is only for future re-activated kits and prevents this happening. 

To fix current data we need to trigger DSM elastic export for the participants listed in the ticket
KitRequestShipping.java 
Noticed issue with full DSM ES export, de-activated records are NOT included and getting exported to ES.
SQL change is to include de-activated record as well during the ES export (as per Whitney's expectation and approval)

Noticed another minor bug in de-activate, re-activate scenario's. Upon re-activate and final scan, we're updating old de-activated record in elastic with new re-activated data . We also insert a new re-activated kit/record in ES.
Made change in KitFinalSentBaseUseCase to skip ES update if de-activated record exists.

QA
Need regression test of de-activate, re-activate scenarios .
Need test of other kit scan scenarios for few studies

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java [hunk 1]
```diff
@@ -91,34 +91,34 @@
...
-                    + "GROUP BY dsm_kit_request_id) groupedKit ON kit.dsm_kit_request_id = groupedKit.dsm_kit_request_id "
-                    + "AND kit.dsm_kit_id = groupedKit.kit_id LEFT JOIN ddp_kit_tracking tracking "
-                    + "ON (kit.kit_label = tracking.kit_label))as wtf) AS kit ON kit.dsm_kit_request_id = request.dsm_kit_request_id "
-                    + "LEFT JOIN ddp_participant_exit ex ON (ex.ddp_instance_id = request.ddp_instance_id "
-                    + "AND ex.ddp_participant_id = request.ddp_participant_id) "
-                    + "LEFT JOIN ddp_kit_request_settings dkc ON (request.ddp_instance_id = dkc.ddp_instance_id "
-                    + "AND request.kit_type_id = dkc.kit_type_id) "
-                    + "LEFT JOIN kit_type ktype ON (request.kit_type_id = ktype.kit_type_id) "
-                    + "WHERE ex.ddp_participant_exit_id is null ";
+            "SELECT * FROM ( SELECT req.upload_reason, kt.kit_type_name, kt.display_name, ddp_site.instance_name, ddp_site.ddp_instance_id, \n" +
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java]: Can you describe the change made here? It is very difficult to see the change given the addition of the newlines?
BTW, what value do the newlines add? Is this query included as part of logging somewhere?

**Review by denniscunningham — COMMENTED** [review]
Best I can tell, you added a field to a couple select queries.
I'm not sure I follow how that applies to your description, since I do not see that field used in any other code.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java (hunk 1 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java]: Dennis: I believe newlines got added when I copy pasted SQL from MySQL workbench. cleaned it up.
Updated SQL to include -deactivated kit/record in the query results along with re-activated kit. Since both point to same kit-request, we're skipping de-activated one in the results . Updated SQL to include it so that de-activated kit/record will be included in ES export

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java [hunk 2]
```diff
@@ -91,34 +91,34 @@
...
                     + "kit.label_url_return, kit.tracking_to_id, "
                     + "kit.tracking_return_id, kit.easypost_tracking_to_url, kit.easypost_tracking_return_url, kit.easypost_to_id, "
                     + "kit.easypost_shipment_status, kit.scan_date, kit.label_date, kit.error, kit.message, "
                     + "kit.receive_date, kit.receive_by, kit.deactivated_date, kit.easypost_address_id_to, kit.deactivation_reason, "
-                    + "tracking.tracking_id,"
-                    + " kit.kit_label, kit.express, kit.test_result, kit.needs_approval, kit.authorization, kit.denial_reason, "
-                    + " kit.authorized_by, kit.ups_tracking_status, kit.ups_return_status, kit.CE_order, kit.collection_date, "
-                    + "  kit.sequencing_restriction, kit.sample_notes FROM ddp_kit kit "
-                    + "INNER JOIN (SELECT dsm_kit_request_id, MAX(dsm_kit_id) AS kit_id FROM ddp_kit "
-                    + "GROUP BY dsm_kit_request_id) groupedKit ON kit.dsm_kit_request_id = groupedKit.dsm_kit_request_id "
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:113]: removed GROUP by dsm_kit_request_id  so that both de-activated and re-activated kits/records will be included in the results

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java [hunk 3]
```diff
@@ -49,13 +49,18 @@ protected void triggerEvents(String ddpLabel, KitRequestShipping kitRequestShipp
         }
         if (kitPayload.getDdpInstanceDto().isESUpdatePossible()) {
             try {
-                UpsertPainlessFacade.of(DBConstants.DDP_KIT_REQUEST_ALIAS, kitRequestShipping,
-                        kitPayload.getDdpInstanceDto(), "ddpLabel", "ddpLabel",
-                        ddpLabel, new PutToNestedScriptBuilder()).export();
+                //to avoid updating old deactivated record in ES because of search by label .. update only if NO de-activated record exists with same label
+                //check if de-activated kit request exists for same kit-request-id and skip ES update
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java:53]: ES is updated anyway with new kitRequestShipping recod in the immediate KitFinalScanUseCase.writeSampleSentToES call during the final-scan

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java [hunk 4]
```diff
@@ -49,13 +49,18 @@ protected void triggerEvents(String ddpLabel, KitRequestShipping kitRequestShipp
...
+                if (kitDao.getDeactivatedKitCountByLabel(kitRequestShipping) == 0) {
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java:54]: Other alternative is check kitId in elastic and skip updating if different kit Id. Went with check in DB

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java (hunk 2 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:113]: I still can't tell what has changed here, so I will just trust that you made the change you intended.
These queries are too long to be understood anyway.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java [hunk 5]
```diff
@@ -28,6 +15,19 @@
...
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.broadinstitute.ddp.db.TransactionWrapper.inTransaction;
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java]: Why was this change necessary?
Typically static imports are before non-statics.
This appears to be generating new checkstyle violations.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java [hunk 6]
```diff
@@ -164,8 +167,8 @@ public class KitDao {
      */
     private static final String SELECT_KIT_TRACKING_BY_KIT_LABEL_OR_TRACKING_ID =
             "select t.kit_label, t.tracking_id, (select email from access_user where user_id = t.scan_by) as scan_by, case "
-            + " when t.scan_date is not null then from_unixtime(t.scan_date/1000) else null end as scan_date from ddp_kit_tracking "
-            + " t where t.kit_label = ? or t.tracking_id = ? limit 1";
+                    + " when t.scan_date is not null then from_unixtime(t.scan_date/1000) else null end as scan_date from ddp_kit_tracking "
+                    + " t where t.kit_label = ? or t.tracking_id = ? limit 1";
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java]: Is this formatting change necessary?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java [hunk 7]
```diff
@@ -502,6 +502,7 @@ private static Map<String, List<KitRequestShipping>> getKitRequestsByKitId(@NonN
                     }
                 }
             } catch (SQLException ex) {
+                ex.printStackTrace();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java]: This typically is not necessary since the exception is logged with the stack trace.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java (hunk 3 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java:53]: Not sure I follow what is going on here.
Won't anything not written to ES here just get written on the next DSM export?

**Review by denniscunningham — COMMENTED** [review]
I am having trouble following your solution.
(First, it is helpful to include the ticket number in the PR description since it will automatically create a link to Jira.)
You say "During the final scan, query the kit type name and update in elastic kitRequestShipping object.". But why or how does this solution work?

I also do not understand this:
"Upon re-activate and final scan, we're updating old de-activated record in elastic with new re-activated data . We also insert a new re-activated kit/record in ES.
Made change in KitFinalSentBaseUseCase to skip ES update if de-activated record exists."
Can you explain what 'old' and 'new' ES KitRequestShipping records are?
Presumably each kit request is represented by one KitRequestShipping record, and that changes state depending on activation status?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java (hunk 3 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalSentBaseUseCase.java:53]: As long as the data in the DB doesn't change, ES exported data shouldn't change. 
Here we're avoiding updating existing kitRequestShipping element/record in this sample participant ES index
https://f7df559584f84f2cab27de822618fb89.us-central1.gcp.cloud.es.io:9243/participants_structured.cmi.cmi-pancan/_search?pretty&q=profile.hruid:PB632F

Avoid updating below existing kitRequestShipping record 
{
                "deactivationReason" : "testDeactivate-1718883724841",
                "ddpLabel" : "N9PPHR8QLIWE4I5",
                "bspCollaboratorSampleId" : "Project_PB632F_SALIVA",
                "dsmKitId" : 59239,
                "deactivatedDate" : 1718883725126,
                "express" : false,
                "error" : false,
                "message" : null,
                "bspCollaboratorParticipantId" : "Project_PB632F",
                "createdBy" : "SYSTEM",
                "ddpInstanceId" : 24,
                "noReturn" : false,
                "kitTypeName" : "SALIVA",
                "ceOrder" : false,
                "dsmKitRequestId" : 59604
              },
              
             with new element/kit created for same label: N9PPHR8QLIWE4I5
             {
                "kitLabel" : "SS_PB632F_SS01",
                "ddpLabel" : "N9PPHR8QLIWE4I5",
                "bspCollaboratorSampleId" : "Project_PB632F_SALIVA",
                "dsmKitId" : 60554,
                "express" : false,
                "error" : true,
                "message" : "To: Connection to the carrier timed out. Please try again later.",
                "bspCollaboratorParticipantId" : "Project_PB632F",
                "scanDate" : 1719889806249,
                "createdBy" : "SYSTEM",
                "trackingReturnId" : "Track_PB632F_SS01",
                "ddpInstanceId" : 24,
                "noReturn" : false,
                "kitTypeName" : "SALIVA",
                "ceOrder" : false,
                "labelDate" : 1719889558467,
                "dsmKitRequestId" : 59604
              },
             
             
             Participant data without the fix
https://f7df559584f84f2cab27de822618fb89.us-central1.gcp.cloud.es.io:9243/participants_structured.cmi.cmi-pancan/_search?pretty&q=profile.hruid:PAVZC3
Search for label: X8ANWLNGOHMZRMO
You can see that both kitRequestShipping elements have same data (new kit info)

**Comment by ssettipalli** [comment]
> I am having trouble following your solution. (First, it is helpful to include the ticket number in the PR description since it will automatically create a link to Jira.) You say "During the final scan, query the kit type name and update in elastic kitRequestShipping object.". But why or how does this solution work?
> "kitTypeName" : is missing in the kitRequestShipping elastic record for re-activated scenarios. Solution is to query from DB and include it in the document while patching elastic index.

> I also do not understand this: "Upon re-activate and final scan, we're updating old de-activated record in elastic with new re-activated data . We also insert a new re-activated kit/record in ES. Made change in KitFinalSentBaseUseCase to skip ES update if de-activated record exists." Can you explain what 'old' and 'new' ES KitRequestShipping records are? Presumably each kit request is represented by one KitRequestShipping record, and that changes state depending on activation status?

Elastic Index has both de-activated record and re-activated record
Ex: search for label: X8ANWLNGOHMZRMO
in https://f7df559584f84f2cab27de822618fb89.us-central1.gcp.cloud.es.io:9243/participants_structured.cmi.cmi-pancan/_search?pretty&q=profile.hruid:PAVZC3
old -  de-activated record [search for "deactivationReason"]
new-  once re-activated and completed final scan record : second record with label: X8ANWLNGOHMZRMO

https://github.com/broadinstitute/ddp-study-server/pull/2906#discussion_r1669153782

I understand its confusing. I am more than happy to get on a quick call to explain the solution and show the scenarios in elastic.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java (hunk 6 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java]: IDE auto formatted and reverted it.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java (hunk 7 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java]: I believe I added for some debug and missed removing it. thank you, cleaned up

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java [hunk 8]
```diff
@@ -492,6 +519,7 @@ public Optional<KitRequestShipping> getKitByDdpLabel(String ddpLabel, String kit
                         kitRequestShipping.setDdpInstanceId(rs.getLong(DBConstants.DDP_INSTANCE_ID));
                         kitRequestShipping.setDdpKitRequestId(rs.getString(DBConstants.DDP_KIT_REQUEST_ID));
                         kitRequestShipping.setKitTypeId(String.valueOf(rs.getInt(DBConstants.KIT_TYPE_ID)));
+                        kitRequestShipping.setKitTypeName(rs.getString(DBConstants.KIT_TYPE_NAME));
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java:522]: Column kit_type_name queried is used in here and setting into KitRequestShipping which will be updated into ES

**Review by denniscunningham — APPROVED** [review]
Sampath and I discussed this PR F2F, and I now better understand the changes.
Sampath is planning some additional testing once this change is on Dev to ensure the changes do not affect other kit operations.
